File: sleeping_barber.py
import threading
import time
import random


customer_ready = threading.Semaphore(0) 
barber_ready = threading.Semaphore(0)
seats_mutex = threading.Lock()

WAITING_ROOM_CAPACITY = 3
waiting_seats = WAITING_ROOM_CAPACITY
TOTAL_CUSTOMERS = 10
shop_is_open = True

def barber():
    global shop_is_open
    print(" [Barber] Shop is open. Waiting for customers...")
    
    while shop_is_open or waiting_seats < WAITING_ROOM_CAPACITY:
        acquired = customer_ready.acquire(timeout=1) 
        
        if not acquired:
            if not shop_is_open and waiting_seats == WAITING_ROOM_CAPACITY:
                break
            continue

        barber_ready.release()
        
        print("  [Barber] Cutting hair...")
        time.sleep(1.5) 
        print(" [Barber] Haircut finished.")

    print(" [Barber] All done. Going home.")

def customer(index):
    global waiting_seats
    
    with seats_mutex:
        if waiting_seats > 0:
            waiting_seats -= 1
            print(f" [Customer {index}] Entered waiting room. Seats left: {waiting_seats}")

            customer_ready.release()
        else:
            print(f" [Customer {index}] Waiting room full. Leaving...")
            return
        
    barber_ready.acquire()
    
    with seats_mutex:
        waiting_seats += 1

if __name__ == "__main__":
    print("--- Simulation Started: Sleeping Barber Problem ---")
    
    barber_thread = threading.Thread(target=barber)
    barber_thread.start()

    customer_threads = []
    for i in range(TOTAL_CUSTOMERS):
        t = threading.Thread(target=customer, args=(i+1,))
        customer_threads.append(t)
        t.start()
        time.sleep(random.uniform(0.5, 1.5))

    for t in customer_threads:
        t.join()

    print(" [Main] No more customers. Closing shop.")
    shop_is_open = False
    barber_thread.join()

    print("--- Simulation Complete. ---")

File: sleeping_barber.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <vector>
#include <chrono>
#include <random>

const int WAITING_ROOM_CAPACITY = 3;
const int TOTAL_CUSTOMERS = 10;

int waiting_seats = WAITING_ROOM_CAPACITY;
bool shop_is_open = true;

std::mutex mtx;
std::condition_variable cv_customer; // Customers waiting
std::condition_variable cv_barber;   // Barber ready
bool barber_is_ready = false;

std::random_device rd;
std::mt19937 gen(rd());
std::uniform_int_distribution<> dis_ms(500, 1500); // ms for customer arrival

void barber() {
    std::cout << " [Barber] Shop is open. Waiting for customers..." << std::endl;

    while (true) {
        std::unique_lock<std::mutex> lock(mtx);

        // Wait until there is a customer or shop is closed
        cv_customer.wait(lock, []() { return waiting_seats < WAITING_ROOM_CAPACITY || !shop_is_open; });

        // If no customers and shop closed, exit
        if (waiting_seats == WAITING_ROOM_CAPACITY && !shop_is_open)
            break;

        // Only serve if there is a customer
        if (waiting_seats < WAITING_ROOM_CAPACITY) {
            barber_is_ready = true;
            cv_barber.notify_one();

            lock.unlock();
            std::cout << "  [Barber] Cutting hair..." << std::endl;
            std::this_thread::sleep_for(std::chrono::milliseconds(1500));
            std::cout << " [Barber] Haircut finished." << std::endl;
        }
    }

    std::cout << " [Barber] All done. Going home." << std::endl;
}

void customer(int index) {
    {
        std::unique_lock<std::mutex> lock(mtx);
        if (waiting_seats > 0) {
            waiting_seats--;
            std::cout << " [Customer " << index << "] Entered waiting room. Seats left: "
                      << waiting_seats << std::endl;
            cv_customer.notify_one();
        } else {
            std::cout << " [Customer " << index << "] Waiting room full. Leaving..." << std::endl;
            return;
        }
    }

    // Wait until barber is ready
    std::unique_lock<std::mutex> lock(mtx);
    cv_barber.wait(lock, []() { return barber_is_ready; });
    barber_is_ready = false;

    // Customer got haircut, free a seat
    waiting_seats++;
}

int main() {
    std::cout << "--- Simulation Started: Sleeping Barber Problem ---" << std::endl;

    std::thread barber_thread(barber);

    std::vector<std::thread> customer_threads;
    for (int i = 1; i <= TOTAL_CUSTOMERS; ++i) {
        customer_threads.emplace_back(customer, i);
        std::this_thread::sleep_for(std::chrono::milliseconds(dis_ms(gen)));
    }

    // Wait for all customers to finish
    for (auto &t : customer_threads)
        t.join();

    // Close the shop and wake barber
    {
        std::lock_guard<std::mutex> lock(mtx);
        shop_is_open = false;
    }
    cv_customer.notify_one();

    barber_thread.join();

    std::cout << "--- Simulation Complete. ---" << std::endl;
    return 0;
}